File: src/main/java/site/gachontable/infra/security/jwt/CustomAccessDeniedHandler.java
package site.gachontable.infra.security.jwt;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;
import site.gachontable.independent.type.ErrorCode;
import site.gachontable.independent.error.ErrorResponse;

import java.io.IOException;

@Component
public class CustomAccessDeniedHandler implements AccessDeniedHandler {

    @Override
    public void handle(HttpServletRequest request,
                       HttpServletResponse response,
                       AccessDeniedException accessDeniedException) throws IOException, ServletException {
        setResponse(response);
    }

    public void setResponse(HttpServletResponse response) throws IOException {
        response.setContentType("application/json;charset=UTF-8");
        response.setStatus(HttpServletResponse.SC_FORBIDDEN);

        ObjectMapper mapper = new ObjectMapper();
        String jsonResponse = mapper.writeValueAsString(
                new ErrorResponse(ErrorCode.ROLE_FORBIDDEN));

        response.getWriter().write(jsonResponse);
    }
}

File: src/main/kotlin/site/gachontable/infra/security/jwt/CustomAccessDeniedHandler.kt
package site.gachontable.infra.security.jwt

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component
import site.gachontable.independent.error.ErrorResponse
import site.gachontable.independent.type.ErrorCode

@Component
class CustomAccessDeniedHandler : AccessDeniedHandler {
    override fun handle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        accessDeniedException: AccessDeniedException,
    ) {
        setResponse(response)
    }

    fun setResponse(response: HttpServletResponse) {
        response.apply {
            contentType = "application/json;charset=UTF-8"
            status = HttpServletResponse.SC_FORBIDDEN
        }

        val mapper = ObjectMapper()
        val jsonResponse = mapper.writeValueAsString(
            ErrorResponse(ErrorCode.ROLE_FORBIDDEN)
        )

        response.writer.write(jsonResponse)
    }
}
